### PR TITLE
feat: adds channel ID label to the Prometheus message_receive_bytes_total and message_send_bytes_total metrics

### LINKED ISSUE #1077: Enhancing Prometheus metrics for BW consumption
As part of https://github.com/celestiaorg/celestia-app/issues/2197, in order to be able to calculate the breakdown of BW consumption per chain ID and reactor, we would like to use to the existing Prometheus metrics in cekestia-core. However, it is not clear whether the current metrics are sufficient and whether we may require additional metrics to accomplish the task. 
This issue intends to track the progress of this effort.

### PR DESCRIPTION
**Pull request by staheri14**
Part of #1077

### CODE REVIEW

File: p2p/metrics.go [hunk 1]
```diff
@@ -87,13 +87,13 @@ func PrometheusMetrics(namespace string, labelsAndValues ...string) *Metrics {
 			Subsystem: MetricsSubsystem,
 			Name:      "message_receive_bytes_total",
 			Help:      "Number of bytes of each message type received.",
-		}, append(labels, "message_type")).With(labelsAndValues...),
+		}, append(labels, "message_type", "chID")).With(labelsAndValues...),
```
Comment (rootulp) [inline p2p/metrics.go:90]: [no change needed][for reviewers] "chID" appears already used many times in this repo. I would've preferred the longer and more precise "channelID" but I don't think it's worth changing b/c we want to minimize the diff to CometBFT.

File: p2p/peer.go [hunk 2]
```diff
@@ -331,7 +331,8 @@ func (p *peer) SendEnvelope(e Envelope) bool {
 	}
 	res := p.Send(e.ChannelID, msgBytes)
 	if res {
-		p.metrics.MessageSendBytesTotal.With("message_type", metricLabelValue).Add(float64(len(msgBytes)))
+		p.metrics.MessageSendBytesTotal.With("message_type",
+			metricLabelValue, "chID", fmt.Sprintf("%#x", e.ChannelID)).Add(float64(len(msgBytes)))
```
Comment (rootulp) [inline p2p/peer.go:335]: [question] Looks like `"%#x"` will print the ChannelID in hexadecimal. Is that the common format used to display ChannelIDs? 

[question] Is there a map from Channel ID => reactor somewhere?

**Review by rootulp — APPROVED** [review]
[question] Since this PR targets v0.34.x-celestia, should it be forward-ported to main? Asking because I though the usual workflow was to target main and backport to v0.34.x-celestia per https://github.com/celestiaorg/celestia-core#branches. cc: @cmwaters 

[question] how was this PR tested? Did you manually verify the metric labels? Do you mind including that in the PR description?

File: p2p/metrics.go (hunk 1 above)
Comment (staheri14) [reply p2p/metrics.go:90]: > [no change needed][for reviewers] "chID" appears already used many times in this repo. I would've preferred the longer and more precise "channelID" 

Agree
> but I don't think it's worth changing b/c we want to minimize the diff to CometBFT.

Agree

File: p2p/peer.go (hunk 2 above)
Comment (staheri14) [reply p2p/peer.go:335]: > [question] Looks like "%#x" will print the ChannelID in hexadecimal. Is that the common format used to display ChannelIDs?

Yes, it is the format used in other Prometheus metrics labels, please see https://github.com/celestiaorg/celestia-core/blob/d344f2712d38e073b06fdf97d4eecf3634fc027e/p2p/peer.go#L528

Comment (staheri14) [reply p2p/peer.go:335]: > [question] Is there a map from Channel ID => reactor somewhere? 

I am not aware of any specs for this, however, every reactor implements the  the `reactor` interface due to which it also supports the  [`GetChannels`](https://github.com/celestiaorg/celestia-core/blob/f7b5c1f8eb71d7dbf8495a3fc69a4a01a2297b53/p2p/base_reactor.go#L23) method which returns the list of supported channel ID for that reactor. Below is an example for the mempool v0 reactor: https://github.com/celestiaorg/celestia-core/blob/3f3b7cc57f5cfc5e846ce781a9a407920e54fb72/mempool/v0/reactor.go#L126

**Review by evan-forbes — APPROVED** [review]
utACK

File: p2p/peer.go (hunk 2 above)
Comment (rootulp) [reply p2p/peer.go:335]: Sounds good, thanks! 

I assume a map from channel ID => reactor name will be useful after this merges but I don't think we want to maintain one so I think this thread can be resolved.

**Comment by staheri14** [comment]
> @staheri14 perhaps its better to create a branch of celestia-app based on v1.x that is using this branch, and then we all view it in action on grafana by updating our mocha node to use that special version of v1.x

@evan-forbes Sounds good to me, will do that and update you here.

**Comment by staheri14** [comment]
> > @staheri14 perhaps its better to create a branch of celestia-app based on v1.x that is using this branch, and then we all view it in action on grafana by updating our mocha node to use that special version of v1.x
> 
> @evan-forbes Sounds good to me, will do that and update you here.

Created [this branch](https://github.com/celestiaorg/celestia-app/tree/sanaz/test-chid) to test it on our mocha node.
cc: @evan-forbes

**Comment by staheri14** [comment]
After some investigation it looks like that chID may have already been attached to the metrics targeted in this PR. So converting this to draft until we verify this.

**Comment by staheri14** [comment]
I was able to run a node (based off the lates main branch and without the changes in this PR) and connect it to mocha-4 and inspect its metrics as provided below. It illustrates that none of the `message_receive_bytes_total` and `message_send_bytes_total` metrics are associated with the channel ID label. In light of this, going to get the PR back to the ready mode.

```
# TYPE cometbft_p2p_message_receive_bytes_total counter
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="blockchain_BlockResponse",version="1.0.0-rc14"} 394149
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="blockchain_StatusResponse",version="1.0.0-rc14"} 7
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_HasVote",version="1.0.0-rc14"} 365
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_NewRoundStep",version="1.0.0-rc14"} 64
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_NewValidBlock",version="1.0.0-rc14"} 50
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="p2p_PexAddrs",version="1.0.0-rc14"} 15121
# HELP cometbft_p2p_message_send_bytes_total Number of bytes of each message type sent.
# TYPE cometbft_p2p_message_send_bytes_total counter
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="blockchain_BlockRequest",version="1.0.0-rc14"} 1900
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="blockchain_StatusResponse",version="1.0.0-rc14"} 7
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="p2p_PexRequest",version="1.0.0-rc14"} 2
```
Below is the full list of metrics, in case needed: 
```
# HELP cometbft_consensus_fast_syncing Whether or not a node is fast syncing. 1 if yes, 0 if no.
# TYPE cometbft_consensus_fast_syncing gauge
cometbft_consensus_fast_syncing{chain_id="mocha-4",version="1.0.0-rc14"} 1
# HELP cometbft_mempool_size Size of the mempool (number of uncommitted transactions).
# TYPE cometbft_mempool_size gauge
cometbft_mempool_size{chain_id="mocha-4",version="1.0.0-rc14"} 0
# HELP cometbft_mempool_successful_txs Number of transactions that successfully made it into a block.
# TYPE cometbft_mempool_successful_txs counter
cometbft_mempool_successful_txs{chain_id="mocha-4",version="1.0.0-rc14"} 0
# HELP cometbft_p2p_message_receive_bytes_total Number of bytes of each message type received.
# TYPE cometbft_p2p_message_receive_bytes_total counter
**cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="blockchain_BlockResponse",version="1.0.0-rc14"} 394149**
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="blockchain_StatusResponse",version="1.0.0-rc14"} 7
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_HasVote",version="1.0.0-rc14"} 365
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_NewRoundStep",version="1.0.0-rc14"} 64
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="consensus_NewValidBlock",version="1.0.0-rc14"} 50
cometbft_p2p_message_receive_bytes_total{chain_id="mocha-4",message_type="p2p_PexAddrs",version="1.0.0-rc14"} 15121
# HELP cometbft_p2p_message_send_bytes_total Number of bytes of each message type sent.
# TYPE cometbft_p2p_message_send_bytes_total counter
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="blockchain_BlockRequest",version="1.0.0-rc14"} 1900
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="blockchain_StatusResponse",version="1.0.0-rc14"} 7
cometbft_p2p_message_send_bytes_total{chain_id="mocha-4",message_type="p2p_PexRequest",version="1.0.0-rc14"} 2
# HELP cometbft_p2p_peer_receive_bytes_total Number of bytes received from a given peer.
# TYPE cometbft_p2p_peer_receive_bytes_total counter
cometbft_p2p_peer_receive_bytes_total{chID="0x0",chain_id="mocha-4",peer_id="34499b1ac473fbb03894c883178ecc83f0d6eaf6",version="1.0.0-rc14"} 15121
cometbft_p2p_peer_receive_bytes_total{chID="0x20",chain_id="mocha-4",peer_id="34499b1ac473fbb03894c883178ecc83f0d6eaf6",version="1.0.0-rc14"} 479
cometbft_p2p_peer_receive_bytes_total{chID="0x40",chain_id="mocha-4",peer_id="34499b1ac473fbb03894c883178ecc83f0d6eaf6",version="1.0.0-rc14"} 394156
# HELP cometbft_p2p_peer_send_bytes_total Number of bytes sent to a given peer.
# TYPE cometbft_p2p_peer_send_bytes_total counter
cometbft_p2p_peer_send_bytes_total{chID="0x0",chain_id="mocha-4",peer_id="34499b1ac473fbb03894c883178ecc83f0d6eaf6",version="1.0.0-rc14"} 2
cometbft_p2p_peer_send_bytes_total{chID="0x40",chain_id="mocha-4",peer_id="34499b1ac473fbb03894c883178ecc83f0d6eaf6",version="1.0.0-rc14"} 1907
# HELP cometbft_p2p_peers Number of peers.
# TYPE cometbft_p2p_peers gauge
cometbft_p2p_peers{chain_id="mocha-4",version="1.0.0-rc14"} 1
# HELP cometbft_state_block_processing_time Time between BeginBlock and EndBlock in ms.
# TYPE cometbft_state_block_processing_time histogram
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="1"} 10
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="11"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="21"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="31"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="41"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="51"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="61"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="71"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="81"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="91"} 14
cometbft_state_block_processing_time_bucket{chain_id="mocha-4",version="1.0.0-rc14",le="+Inf"} 14
cometbft_state_block_processing_time_sum{chain_id="mocha-4",version="1.0.0-rc14"} 21.592999999999996
cometbft_state_block_processing_time_count{chain_id="mocha-4",version="1.0.0-rc14"} 14
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 2.0124e-05
go_gc_duration_seconds{quantile="0.25"} 2.8126e-05
go_gc_duration_seconds{quantile="0.5"} 4.5291e-05
go_gc_duration_seconds{quantile="0.75"} 5.6333e-05
go_gc_duration_seconds{quantile="1"} 0.000195125
go_gc_duration_seconds_sum 0.000890792
go_gc_duration_seconds_count 13
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 671
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.21.0"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 1.6826864e+08
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 6.45035184e+08
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.498264e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 1.80352e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 1.033076e+07
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 1.6826864e+08
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.2939264e+08
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 1.76922624e+08
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 1.529264e+06
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.398528e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 3.06315264e+08
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.694115731110122e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 3.332784e+06
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 9600
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 15600
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 1.42548e+06
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.434048e+06
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 2.24743112e+08
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.61164e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 4.063232e+06
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 4.063232e+06
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 3.25268808e+08
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 14
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# TYPE promhttp_metric_handler_requests_in_flight gauge
promhttp_metric_handler_requests_in_flight 1
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
# TYPE promhttp_metric_handler_requests_total counter
promhttp_metric_handler_requests_total{code="200"} 0
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
```

File: p2p/peer.go [hunk 3]
```diff
@@ -532,7 +534,8 @@ func createMConnection(
 			}
 		}
 		p.metrics.PeerReceiveBytesTotal.With(labels...).Add(float64(len(msgBytes)))
-		p.metrics.MessageReceiveBytesTotal.With("message_type", p.mlc.ValueToMetricLabel(msg)).Add(float64(len(msgBytes)))
+		p.metrics.MessageReceiveBytesTotal.With("message_type",
+			p.mlc.ValueToMetricLabel(msg), "chID", fmt.Sprintf("%#x", chID)).Add(float64(len(msgBytes)))
```
Comment (evan-forbes) [inline p2p/peer.go:538]: [no changes needed]
using fmt.Sprintf in a heavy path is something to keep an eye on in the future

https://github.com/tendermint/tendermint/pull/8845#issuecomment-1164483127

**Review by evan-forbes — APPROVED** [review]
I know the chID doesn't appear in the metrics w/o this change, but did we ever figure out why we are able to index by channel w/o it?

**Comment by staheri14** [comment]
>  did we ever figure out why we are able to index by channel w/o it?

Can you please elaborate on this? When were we able to do index by channel id? And for which metrics?

**Comment by evan-forbes** [comment]
ahhh I'm dumb ok this is from the peer_send_bytes_total and peer_receive_bytes_total, which is indexed by channel.

Screenshot from 2023-09-06 18-38-59

**Comment by staheri14** [comment]
If we would like to have more granularity on the bw consumption per Peer, we can use `peer_send_bytes_total` and `peer_receive_bytes_total`. I'll see if we can further agreegate those.
Nevertheless, given that we now have `message_receive_bytes_total` and `message_send_bytes_total` metrics with channel ID labels available, I propose to proceed as follows:
- Compute per peer data transmission rate using `peer_send_bytes_total` and `peer_receive_bytes_total`
- Compute total  data transmission rate using `message_receive_bytes_total` and `message_send_bytes_total`.
- We can additionally and optionally cross check the two (the aggregate of the data transmission rate of all the peers should be equal to the the data transmission rate of `message_receive_bytes_total` and `message_send_bytes_total`) . 

If there is no opposing opinion, I am going to start making queries and gathering the results. 
cc: @evan-forbes

**Review by evan-forbes — APPROVED** [review]
sgtm!

**Comment by staheri14** [comment]
@evan-forbes Shall we merge this (or do we want to keep it open for future metrics that we may come up with)? the channel ID proved useful in our last traffic rate report using Prometheus metrics. I am more inclined toward merging it, wdyt?

**Comment by staheri14** [comment]
Got @evan-forbes confirmation in a sync call, so going to merge it.

**Pull request merged**